cdm/misc/src/main/java/ucar/nc2/geotiff/GeotiffWriter.java

```diff
@@ -213,17 +300,54 @@ private void writeMetadata(boolean greyScale, double xStart, double yStart, doub
+      } else {
+        geotiff.addTag(new IFDEntry(Tag.PhotometricInterpretation, FieldType.SHORT).setValue(1)); // black is zero :
+                                                                                                  // not used?
+      }
       // standard tags for SampleFormat ( see TIFF spec, section 19)
-      geotiff.addTag(new IFDEntry(Tag.BitsPerSample, FieldType.SHORT).setValue(32)); // 32 bits per sample
-      geotiff.addTag(new IFDEntry(Tag.SampleFormat, FieldType.SHORT).setValue(3)); // Sample Format
+      geotiff.addTag(new IFDEntry(Tag.BitsPerSample, FieldType.SHORT).setValue(elemSize * 8));
+
+      if (dtype.isIntegral()) {
+        geotiff.addTag(new IFDEntry(Tag.SampleFormat, FieldType.SHORT).setValue(dtype.isUnsigned() ? 1 : 2)); // UINT or
+                                                                                                              // INT
+      } else if (dtype.isFloatingPoint()) {
+        geotiff.addTag(new IFDEntry(Tag.SampleFormat, FieldType.SHORT).setValue(3)); // IEEE Floating Point Type
+      } else {
+        throw new IllegalArgumentException("Unsupported data type for geotiff: " + dtype);
+      }
       geotiff.addTag(new IFDEntry(Tag.SamplesPerPixel, FieldType.SHORT).setValue(1));
-      float min = (float) (dataMinMax.min);
-      float max = (float) (dataMinMax.max);
-      geotiff.addTag(new IFDEntry(Tag.SMinSampleValue, FieldType.FLOAT).setValue(min));
-      geotiff.addTag(new IFDEntry(Tag.SMaxSampleValue, FieldType.FLOAT).setValue(max));
-      geotiff.addTag(new IFDEntry(Tag.GDALNoData, FieldType.FLOAT).setValue(min - 1.f));
+
+      if (dtype.isFloatingPoint()) {
+        float min = (float) (dataMinMax.min);
+        float max = (float) (dataMinMax.max);
+        geotiff.addTag(new IFDEntry(Tag.SMinSampleValue, FieldType.FLOAT).setValue(min));
+        geotiff.addTag(new IFDEntry(Tag.SMaxSampleValue, FieldType.FLOAT).setValue(max));
+        geotiff.addTag(new IFDEntry(Tag.GDALNoData, FieldType.FLOAT).setValue(min - 1.f));
+      } else if (dtype.isIntegral()) {
+        int min = (int) (dataMinMax.min);
+        int max = (int) (dataMinMax.max);
+        FieldType ftype;
+        DataType sdtype = dtype.withSignedness(DataType.Signedness.SIGNED);
+        if (sdtype == DataType.BYTE) {
+          ftype = FieldType.BYTE;
+        } else if (sdtype == DataType.SHORT) {
+          ftype = FieldType.SHORT;
+        } else if (sdtype == DataType.INT) {
+          // A geotiff LONG is really a 4-byte regular integer
+          ftype = FieldType.LONG;
+        } else {
+          throw new IllegalArgumentException("Unsupported dtype: " + dtype + " sdtype: " + sdtype);
+        }
+        geotiff.addTag(new IFDEntry(Tag.SMinSampleValue, ftype).setValue(min));
+        geotiff.addTag(new IFDEntry(Tag.SMaxSampleValue, ftype).setValue(max));
+        geotiff.addTag(new IFDEntry(Tag.GDALNoData, ftype).setValue(min - 1));
```

Comment:
Just discovered an odd bug here. When the original data's range goes from 0 to 255 for ubyte type, this attribute gets set to 0, which messes up the color table since I wanted 255 to be NoData. At the moment, my guess is that a normalization elsewhere is making it that 1 is always the minimum and therefore 0 always becomes NoData. So, I might need to double-check whether or not normalization is even desired in the first place and allow other mechanisms for specifying a NoData attribute.

Reply:
Alright, I think I understand what the bug is. Over at https://github.com/Unidata/netcdf-java/pull/1066/files#diff-b37a37fa215c4b41463e379358f9e9748244261adee3067e72fea3e231344aeaR851, when not writing out greyscale images, I made it so that writing out integer, short, or byte data is done without scaling and missing data handling. So, I shouldn't be setting this field here with "min - 1" since scaling of the original data isn't happening. Instead, I probably should be checking the `grid` object to find out what its missing data setting is if any and use that, maybe?

Reply:
Hmmm, seems like the underlaying VariableDS object returns a list of missing values rather than a single definitive one. Is there an accepted approach to determine a dataset's "missing value"? In xarray & python-netcdf4, it seems like it is usually set as the dataset's "_FillValue" attribute, but I have also seen it set as the "missing" attribute.

Reply:
A variable can have multiple missing values, in addition to a fill value, If your data only has one missing values, `getMissing` will return an array of just that value.

Reply:
Right, I'm just trying to figure out what should be considered "the right/best way" since I can only encode a single NoData value. Like, are the values in the list ordered by "priority" or something? Maybe it would be best to not guess at the user's intent and not encode this at all?